File: rust/koru/examples/hello.rs
// SPDX-License-Identifier: MIT

//! Braam's `examples/hello/hello.cpp`, in Rust. Its C++ original is
//! seventeen lines and names no executor; so does this.

use koru::{Args, Result, write_all};

#[koru::main]
async fn main(args: Args) -> Result<i32> {
    let mut who = "world";
    if args.size() > 1 {
        who = &args[1];
    }

    write_all(koru::stdout(), "Hello, ").await?;
    write_all(koru::stdout(), who).await?;
    write_all(koru::stdout(), "!\n").await?;

    Ok(0)
}

File: cpp/examples/hello.cpp
// SPDX-License-Identifier: MIT
//
// T44's done test: Braam's `examples/hello/hello.cpp`, unchanged but for the
// include line. It names no executor, no ring, no namespace and no header but
// this one, and `rust/runtime/examples/hello.rs` is the same program in the
// other binding — scripts/cpp.sh runs both and compares the bytes.

#include <koru/braam.hpp>

Task<Result<i32>> koru_main(Args args)
{
    Str who = "world";
    if (args.size() > 1)
        who = args[1];

    CO_TRY_VOID(co_await write_all(out_fd(), "Hello, "));
    CO_TRY_VOID(co_await write_all(out_fd(), who));
    CO_TRY_VOID(co_await write_all(out_fd(), "!\n"));

    co_return 0;
}
